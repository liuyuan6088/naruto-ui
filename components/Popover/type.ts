
import { PlacementType, ICommonSubgroupProps } from '../Portal/type'

export interface IPopoverProps extends ICommonSubgroupProps {
  // 弹出方向
  placement?: PlacementType;
  // 卡片标题
  title?: React.ReactNode;
  // 卡片内容
  content?: React.ReactNode;
}