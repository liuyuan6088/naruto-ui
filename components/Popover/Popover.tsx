import * as React from 'react'
import Portal from '../Portal'
import { IPopoverProps } from './type'
import { primaryName } from '../utils/constant'
import './style/index.less'

const defaultProps: IPopoverProps = {
  prefixCls: `${primaryName}-popover`,
  placement: 'top',
  offset: 2,
  trigger: 'hover',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1
}

const Popover: React.FC<IPopoverProps> = props => {
  const {
    prefixCls,
    visible,
    trigger,
    placement,
    offset,
    onVisibleChange,
    title,
    content: propsContent,
    children,
    ...rest
  } = props

  const content = (
    <div className={`${prefixCls}-content-inner`}>
      <div className={`${prefixCls}-content-title`}>{title}</div>
      <div className={`${prefixCls}-content-info`}>{propsContent}</div>
    </div>
  )

  return (
    <Portal
      prefixCls={prefixCls}
      visible={visible}
      content={content}
      transitionName="zoom-big"
      offset={offset}
      placement={placement}
      trigger={trigger}
      onVisibleChange={onVisibleChange}
      wrapperClassName={`${prefixCls}-wrapper`}
      hasTriangle
      {...rest}
      mode="popover"
    >
      {children}
    </Portal>
  )
}

Popover.defaultProps = defaultProps

export default Popover
