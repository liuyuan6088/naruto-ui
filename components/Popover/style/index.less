@import '../../styles/theme/default.less';

@popover-prefix-cls: ~"@{primary-name}-popover";

.@{popover-prefix-cls} {
  position: absolute;
  z-index: @popover-basic-zIndex;
  font-size: 14px;
  font-variant: tabular-nums;
  line-height: 1.5;
  list-style: none;
  cursor: pointer;

  &-wrapper {
    display: inline-block;
    cursor: pointer;
  }

  &-content {
    background-color: #fff;
    border-radius: @bordor-radius;
    box-shadow: @popover-content-shadow;
    text-align: left;
    text-decoration: none;
    word-wrap: break-word;
    position: relative;
    cursor: pointer;

    &-title {
      padding: 5px 15px;
      min-width: 177px;
      min-height: 32px;
      margin: 0;
      padding: 5px 16px 4px;
      color: @gray-2;
      font-weight: 600;
      letter-spacing: 1px;
      border-bottom: 1px solid @border-color-0;
    }

    &-info {
      padding: 12px 16px;
      margin: 0;
      color: @gray-0;
    }
  }

  &-top {
    transform-origin: center 100%;
  }
  &-topLeft {
    transform-origin: 0 100%;
  }
  &-topRight {
    transform-origin: 100% 100%;
  }
  &-bottom {
    transform-origin: center 0;
  }
  &-bottomLeft {
    transform-origin: 0 0;
  }
  &-bottomRight {
    transform-origin: 100% 0;
  }
  &-left {
    transform-origin: 100% center;
  }
  &-leftTop {
    transform-origin: 100% 0;
  }
  &-leftBottom {
    transform-origin: 100% 100%;
  }
  &-right {
    transform-origin: 0 center;
  }
  &-rightTop {
    transform-origin: 0 0;
  }
  &-rightBottom {
    transform-origin: 0 100%;
  }

  &-triangle {
    .@{popover-prefix-cls}-content {
      &::after {
        content: '';
        position: absolute;
        width: 0;
        height: 0;
        border-color: transparent;
        border-style: solid;
        box-sizing: border-box;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-top,
  &-triangle.@{popover-prefix-cls}-topLeft,
  &-triangle.@{popover-prefix-cls}-topRight {
    padding-bottom: 10px;
    .@{popover-prefix-cls}-content {
      &::after {
        bottom: -5px;
        border-width: 5px 5px 0;
        border-top-color: #fff;
      }
    }
  }
  
  &-triangle.@{popover-prefix-cls}-bottom,
  &-triangle.@{popover-prefix-cls}-bottomLeft,
  &-triangle.@{popover-prefix-cls}-bottomRight {
    padding-top: 10px;
    .@{popover-prefix-cls}-content {
      &::after {
        top: -5px;
        border-width: 0 5px 5px;
        border-bottom-color: #fff;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-left,
  &-triangle.@{popover-prefix-cls}-leftTop,
  &-triangle.@{popover-prefix-cls}-leftBottom {
    padding-right: 10px;
    .@{popover-prefix-cls}-content {
      &::after {
        right: -5px;
        border-width: 5px 0 5px 5px;
        border-left-color: #fff;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-right,
  &-triangle.@{popover-prefix-cls}-rightTop,
  &-triangle.@{popover-prefix-cls}-rightBottom {
    padding-left: 10px;
    .@{popover-prefix-cls}-content {
      &::after {
        left: -5px;
        border-width: 5px 5px 5px 0;
        border-right-color: #fff;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-top,
  &-triangle.@{popover-prefix-cls}-bottom {
    .@{popover-prefix-cls}-content {
      &::after {
        left: 50%;
        transform: translateX(-50%);
      }
    }
  }
  &-triangle.@{popover-prefix-cls}-topLeft,
  &-triangle.@{popover-prefix-cls}-bottomLeft {
    .@{popover-prefix-cls}-content {
      &::after {
        left: 16px;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-topRight,
  &-triangle.@{popover-prefix-cls}-bottomRight {
    .@{popover-prefix-cls}-content {
      &::after {
        right: 16px;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-leftTop,
  &-triangle.@{popover-prefix-cls}-rightTop {
    .@{popover-prefix-cls}-content {
      &::after {
        top: 10px;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-left,
  &-triangle.@{popover-prefix-cls}-right {
    .@{popover-prefix-cls}-content {
      &::after {
        top: 50%;
        margin-top: -5px;
      }
    }
  }

  &-triangle.@{popover-prefix-cls}-leftBottom,
  &-triangle.@{popover-prefix-cls}-rightBottom {
    .@{popover-prefix-cls}-content {
      &::after {
        bottom: 10px;
      }
    }
  }
}