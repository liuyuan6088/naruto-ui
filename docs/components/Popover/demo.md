---
imports:
  import Demo from './demo.tsx';
  import DemoCode from '!raw-loader!./demo.tsx';

---

## Popover 气泡卡片

点击/鼠标移入元素，弹出气泡式的卡片浮层。

## 使用Popover

:::demo
  <CodeBox code={DemoCode}>
    <Demo />
  </CodeBox>
:::

## Props
| 参数     | 说明     | 类型   | 可选值 | 默认值 |
| -------- | -------- | ------ | ------ | ------ |
| title | 标题 | string / React.ReactNode | - | - |
| content | 内容 | string / React.ReactNode | - | - |
| placement | 弹出方向 | PlacementType | - | top |
| offset | 偏移距离 | number | - | - |
| onVisibleChange | 显隐状态变化的回调 | (visible: boolean) => void | - | - |
| visible | 是否可见 受控 | boolean | - | - |
| trigger | 触发 dropdown 的方式 | Trigger | - | hover |
| mouseEnterDelay | 鼠标移入后延时多少才显示 Tooltip 单位：秒 | number | - | - |
| mouseLeaveDelay | 鼠标移出后延时多少才隐藏 Tooltip 单位：秒 | number | - | - |
| dropDownStyle | dropDown 样式 | React.CSSProperties | - | - |
| dropDownClassName | dropDown 类名 | string | - | - |

