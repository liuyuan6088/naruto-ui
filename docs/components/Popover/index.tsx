import * as React from 'react';
// @ts-ignore
import Demo from './demo.md';

export default function () {
  return <Demo />
}
