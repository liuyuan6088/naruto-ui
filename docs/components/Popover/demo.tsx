import * as React from 'react';
import { Popover, Button } from 'components';

const topList = ['topLeft', 'top', 'topRight']
const leftList = ['leftTop', 'left', 'leftBottom']
const bottomList = ['bottomLeft', 'bottom', 'bottomRight']
const rightList = ['rightTop', 'right', 'rightBottom']

const Demo: React.FC = () => {

  const content = (
    <div>
      <div>content!!!!content!!!!</div>
      <div>content!!!!content!!!!</div>
      <div>content!!!!content!!!!</div>
      <div>content!!!!content!!!!</div>
    </div>
  )

  return (
    <React.Fragment>

      <p>基本用法</p>
      <div className='demo-spin-1'>
        <Popover title='标题' content={content}>
          <Button>基本用法</Button>
        </Popover>
      </div>

      <p>两种触发方式</p>
      <div className='demo-spin-1'>
        <Popover title='标题' content={content} trigger='click'>
          <Button>click</Button>
        </Popover>

        <Popover title='标题' content={content}>
          <Button>hover</Button>
        </Popover>
      </div>

      <p>placement位置有 12 个方向</p>
      <div className='demo-tooltip-1'>
        {
          [topList, leftList, bottomList, rightList].map((e) => (
            <div key={e[1]} className={`demo-tooltip-${e[1]}`}>
              {
                e.map((ee) => (
                  <Popover key={ee} placement={(ee as any)} title='标题' content={content}>
                    <Button>{ee}</Button>
                  </Popover>
                ))
              }
            </div>
          ))
        }
      </div>

    </React.Fragment>
  )
}

export default Demo;
