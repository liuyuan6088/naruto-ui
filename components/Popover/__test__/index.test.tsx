import * as React from 'react'
import { mount } from 'enzyme'
import Popover from '../index'

describe('Popover', () => {
  it('render Popover', () => {
    const wrapper = mount(
      <Popover title="" content="" transitionName="" mouseEnterDelay={0} mouseLeaveDelay={0}>
        <div id="hello">Hello world!</div>
      </Popover>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
