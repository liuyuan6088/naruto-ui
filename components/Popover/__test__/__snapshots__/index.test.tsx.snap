// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover render Popover 1`] = `ReactWrapper {}`;
